File: ios/Cleanup/Core/Services/ArchiveAuthService.swift
import Foundation
import AuthenticationServices

actor ArchiveAuthService {
    static let shared = ArchiveAuthService()

    private let cognitoIdentityPoolId = "us-east-1:3f2f14d8-55f1-4998-9f87-e995528d6710"
    private let region = "us-east-1"

    private let identityIdKey = "archive_cognito_identity_id"
    private let accessKeyIdKey = "archive_access_key_id"
    private let secretKeyKey = "archive_secret_key"
    private let sessionTokenKey = "archive_session_token"
    private let tokenExpiryKey = "archive_token_expiry"
    private let appleTokenKey = "archive_apple_id_token"
    private let registeredKey = "archive_backend_registered"

    private(set) var identityId: String?
    private(set) var accessKeyId: String?
    private(set) var secretKey: String?
    private(set) var sessionToken: String?
    private var tokenExpiresAt: Date?
    private var cachedAppleToken: String?
    private var isBackendRegistered: Bool

    struct AWSCredentials {
        let accessKeyId: String
        let secretKey: String
        let sessionToken: String
        let region: String
    }

    private init() {
        // Restore persisted auth state
        identityId = UserDefaults.standard.string(forKey: identityIdKey)
        accessKeyId = UserDefaults.standard.string(forKey: accessKeyIdKey)
        secretKey = UserDefaults.standard.string(forKey: secretKeyKey)
        sessionToken = UserDefaults.standard.string(forKey: sessionTokenKey)
        cachedAppleToken = UserDefaults.standard.string(forKey: appleTokenKey)
        isBackendRegistered = UserDefaults.standard.bool(forKey: registeredKey)
        if let expiry = UserDefaults.standard.object(forKey: tokenExpiryKey) as? Date {
            tokenExpiresAt = expiry
        }
    }

    var isAuthenticated: Bool {
        identityId != nil && accessKeyId != nil && secretKey != nil && sessionToken != nil &&
        (tokenExpiresAt ?? .distantPast) > Date()
    }

    /// True only when the user is both authenticated AND registered with the backend.
    var isFullySetUp: Bool {
        isAuthenticated && isBackendRegistered
    }

    // MARK: - Sign In with Apple → Cognito

    func signIn(appleIdToken: String) async throws {
        // 1. Exchange Apple ID token for Cognito identity
        let cognitoIdentity = try await getCognitoIdentity(appleToken: appleIdToken)

        // 2. Get temporary AWS credentials
        let credentials = try await getCognitoCredentials(
            identityId: cognitoIdentity.identityId,
            appleToken: appleIdToken
        )

        // 3. Store all credentials
        self.identityId = cognitoIdentity.identityId
        self.accessKeyId = credentials.accessKeyId
        self.secretKey = credentials.secretKey
        self.sessionToken = credentials.sessionToken
        self.tokenExpiresAt = credentials.expiresAt
        self.cachedAppleToken = appleIdToken

        // 4. Persist
        persistCredentials()

        // 5. Register with backend (using SigV4-signed request)
        try await registerWithBackend()
        isBackendRegistered = true
        UserDefaults.standard.set(true, forKey: registeredKey)
    }

    /// Ensures the user is registered with the backend.
    /// Call this before making API calls to handle the case where sign-in partially succeeded.
    func ensureRegistered() async throws {
        guard isAuthenticated else {
            throw ArchiveError.notAuthenticated
        }
        if isBackendRegistered { return }
        try await refreshCredentialsIfNeeded()
        try await registerWithBackend()
        isBackendRegistered = true
        UserDefaults.standard.set(true, forKey: registeredKey)
    }

    func getCredentials() async throws -> AWSCredentials {
        try await refreshCredentialsIfNeeded()
        guard let ak = accessKeyId, let sk = secretKey, let st = sessionToken else {
            throw ArchiveError.notAuthenticated
        }
        return AWSCredentials(accessKeyId: ak, secretKey: sk, sessionToken: st, region: region)
    }

    func refreshCredentialsIfNeeded() async throws {
        guard let expiresAt = tokenExpiresAt else {
            throw ArchiveError.notAuthenticated
        }
        // Refresh if expiring within 5 minutes
        guard expiresAt.timeIntervalSinceNow < 300 else { return }

        guard let identity = identityId else {
            throw ArchiveError.notAuthenticated
        }

        let credentials = try await getCognitoCredentials(
            identityId: identity,
            appleToken: cachedAppleToken
        )
        self.accessKeyId = credentials.accessKeyId
        self.secretKey = credentials.secretKey
        self.sessionToken = credentials.sessionToken
        self.tokenExpiresAt = credentials.expiresAt

        persistCredentials()
    }

    func signOut() {
        identityId = nil
        accessKeyId = nil
        secretKey = nil
        sessionToken = nil
        tokenExpiresAt = nil
        cachedAppleToken = nil
        isBackendRegistered = false
        for key in [identityIdKey, accessKeyIdKey, secretKeyKey, sessionTokenKey, tokenExpiryKey, appleTokenKey, registeredKey] {
            UserDefaults.standard.removeObject(forKey: key)
        }
    }

    // MARK: - Private

    private func persistCredentials() {
        UserDefaults.standard.set(identityId, forKey: identityIdKey)
        UserDefaults.standard.set(accessKeyId, forKey: accessKeyIdKey)
        UserDefaults.standard.set(secretKey, forKey: secretKeyKey)
        UserDefaults.standard.set(sessionToken, forKey: sessionTokenKey)
        UserDefaults.standard.set(tokenExpiresAt, forKey: tokenExpiryKey)
        UserDefaults.standard.set(cachedAppleToken, forKey: appleTokenKey)
    }

    // MARK: - Cognito API Calls

    private struct CognitoIdentityResponse {
        let identityId: String
    }

    private struct CognitoCredentialsResponse {
        let accessKeyId: String
        let secretKey: String
        let sessionToken: String
        let expiresAt: Date
    }

    private func getCognitoIdentity(appleToken: String) async throws -> CognitoIdentityResponse {
        let url = URL(string: "https://cognito-identity.\(region).amazonaws.com")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-amz-json-1.1", forHTTPHeaderField: "Content-Type")
        request.setValue("AWSCognitoIdentityService.GetId", forHTTPHeaderField: "X-Amz-Target")

        let body: [String: Any] = [
            "IdentityPoolId": cognitoIdentityPoolId,
            "Logins": [
                "appleid.apple.com": appleToken
            ]
        ]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ArchiveError.invalidResponse
        }
        guard (200...299).contains(httpResponse.statusCode) else {
            throw ArchiveError.serverError(httpResponse.statusCode)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let identityId = json["IdentityId"] as? String else {
            throw ArchiveError.invalidResponse
        }

        return CognitoIdentityResponse(identityId: identityId)
    }

    private func getCognitoCredentials(identityId: String, appleToken: String?) async throws -> CognitoCredentialsResponse {
        let url = URL(string: "https://cognito-identity.\(region).amazonaws.com")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-amz-json-1.1", forHTTPHeaderField: "Content-Type")
        request.setValue("AWSCognitoIdentityService.GetCredentialsForIdentity", forHTTPHeaderField: "X-Amz-Target")

        var body: [String: Any] = ["IdentityId": identityId]
        if let token = appleToken {
            body["Logins"] = ["appleid.apple.com": token]
        }
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ArchiveError.invalidResponse
        }
        guard (200...299).contains(httpResponse.statusCode) else {
            throw ArchiveError.serverError(httpResponse.statusCode)
        }

        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let credentials = json["Credentials"] as? [String: Any],
              let accessKeyId = credentials["AccessKeyId"] as? String,
              let secretKey = credentials["SecretKey"] as? String,
              let sessionToken = credentials["SessionToken"] as? String,
              let expiration = credentials["Expiration"] as? TimeInterval else {
            throw ArchiveError.invalidResponse
        }

        return CognitoCredentialsResponse(
            accessKeyId: accessKeyId,
            secretKey: secretKey,
            sessionToken: sessionToken,
            expiresAt: Date(timeIntervalSince1970: expiration)
        )
    }

    /// Syncs the current archive subscription to the backend so quota is set correctly.
    func syncSubscription() async throws {
        guard isAuthenticated else { return }
        try await refreshCredentialsIfNeeded()
        try await registerWithBackend()
    }

    private func registerWithBackend() async throws {
        let creds = try await getCredentials()
        guard let identity = identityId else {
            throw ArchiveError.notAuthenticated
        }
        guard let url = URL(string: "\(ArchiveAPIClient.baseURL)/auth/register") else {
            throw ArchiveError.invalidResponse
        }

        // Include subscription product ID so backend can set quota
        let subscriptionProductId = await MainActor.run {
            EntitlementManager.shared.currentArchiveTier?.rawValue
        }

        var bodyDict: [String: Any] = ["source": "ios"]
        if let productId = subscriptionProductId {
            bodyDict["subscriptionProductId"] = productId
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(identity, forHTTPHeaderField: "X-Identity-Id")
        request.httpBody = try JSONSerialization.data(withJSONObject: bodyDict)

        let signedRequest = SigV4Signer.sign(
            request: request,
            credentials: creds,
            service: "execute-api"
        )

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: signedRequest)
        } catch {
            throw ArchiveError.uploadFailed("Network error: \(error.localizedDescription)")
        }

        guard let httpResponse = response as? HTTPURLResponse,
              (200...299).contains(httpResponse.statusCode) else {
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
            let body = String(data: data, encoding: .utf8) ?? "No body"
            throw ArchiveError.uploadFailed("Register HTTP \(statusCode): \(body)")
        }
    }
}

File: android/app/src/main/java/com/kreativekoala/cleanup/domain/service/ArchiveAuthService.kt
package com.kreativekoala.cleanup.domain.service

import android.content.Context
import com.kreativekoala.cleanup.data.remote.SigV4Interceptor
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.Date
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Port of iOS ArchiveAuthService.
 *
 * Handles Google Sign-In → AWS Cognito Identity Pool → temporary AWS credentials.
 * Used for authenticating archive API requests.
 */
@Singleton
class ArchiveAuthService @Inject constructor(
    @ApplicationContext private val context: Context
) {
    companion object {
        private const val COGNITO_IDENTITY_POOL_ID = "us-east-1:3f2f14d8-55f1-4998-9f87-e995528d6710"
        private const val REGION = "us-east-1"
        private const val COGNITO_ENDPOINT = "https://cognito-identity.us-east-1.amazonaws.com"

        private const val PREFS_NAME = "archive_auth_prefs"
        private const val KEY_IDENTITY_ID = "identity_id"
        private const val KEY_ACCESS_KEY_ID = "access_key_id"
        private const val KEY_SECRET_KEY = "secret_key"
        private const val KEY_SESSION_TOKEN = "session_token"
        private const val KEY_TOKEN_EXPIRY = "token_expiry"
        private const val KEY_GOOGLE_TOKEN = "google_id_token"
        private const val KEY_BACKEND_REGISTERED = "backend_registered"
    }

    private val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    private var identityId: String? = prefs.getString(KEY_IDENTITY_ID, null)
    private var accessKeyId: String? = prefs.getString(KEY_ACCESS_KEY_ID, null)
    private var secretKey: String? = prefs.getString(KEY_SECRET_KEY, null)
    private var sessionToken: String? = prefs.getString(KEY_SESSION_TOKEN, null)
    private var tokenExpiresAt: Long = prefs.getLong(KEY_TOKEN_EXPIRY, 0)
    private var cachedGoogleToken: String? = prefs.getString(KEY_GOOGLE_TOKEN, null)
    private var isBackendRegistered: Boolean = prefs.getBoolean(KEY_BACKEND_REGISTERED, false)

    private val _isAuthenticated = MutableStateFlow(checkAuthenticated())
    val isAuthenticated: StateFlow<Boolean> = _isAuthenticated.asStateFlow()

    private val _isFullySetUp = MutableStateFlow(checkAuthenticated() && isBackendRegistered)
    val isFullySetUp: StateFlow<Boolean> = _isFullySetUp.asStateFlow()

    private fun checkAuthenticated(): Boolean {
        return identityId != null && accessKeyId != null &&
                secretKey != null && sessionToken != null &&
                tokenExpiresAt > System.currentTimeMillis()
    }

    // MARK: - Sign In with Google → Cognito

    suspend fun signIn(googleIdToken: String) = withContext(Dispatchers.IO) {
        // 1. Exchange Google ID token for Cognito identity
        val cognitoIdentityId = getCognitoIdentity(googleIdToken)

        // 2. Get temporary AWS credentials
        val credentials = getCognitoCredentials(cognitoIdentityId, googleIdToken)

        // 3. Store credentials
        identityId = cognitoIdentityId
        accessKeyId = credentials.accessKeyId
        secretKey = credentials.secretKey
        sessionToken = credentials.sessionToken
        tokenExpiresAt = credentials.expiresAtMillis
        cachedGoogleToken = googleIdToken

        persistCredentials()
        _isAuthenticated.value = true

        // 4. Register with backend
        registerWithBackend()
        isBackendRegistered = true
        prefs.edit().putBoolean(KEY_BACKEND_REGISTERED, true).apply()
        _isFullySetUp.value = true
    }

    suspend fun ensureRegistered() = withContext(Dispatchers.IO) {
        if (!checkAuthenticated()) throw ArchiveError.NotAuthenticated
        if (isBackendRegistered) return@withContext
        refreshCredentialsIfNeeded()
        registerWithBackend()
        isBackendRegistered = true
        prefs.edit().putBoolean(KEY_BACKEND_REGISTERED, true).apply()
        _isFullySetUp.value = true
    }

    fun getCredentials(): SigV4Interceptor.AwsCredentials? {
        val ak = accessKeyId ?: return null
        val sk = secretKey ?: return null
        val st = sessionToken ?: return null
        val id = identityId ?: return null
        if (tokenExpiresAt <= System.currentTimeMillis()) return null
        return SigV4Interceptor.AwsCredentials(
            accessKeyId = ak,
            secretKey = sk,
            sessionToken = st,
            identityId = id,
            region = REGION
        )
    }

    suspend fun refreshCredentialsIfNeeded() = withContext(Dispatchers.IO) {
        // Refresh if expiring within 5 minutes
        if (tokenExpiresAt - System.currentTimeMillis() > 5 * 60 * 1000) return@withContext

        val identity = identityId ?: throw ArchiveError.NotAuthenticated
        val credentials = getCognitoCredentials(identity, cachedGoogleToken)

        accessKeyId = credentials.accessKeyId
        secretKey = credentials.secretKey
        sessionToken = credentials.sessionToken
        tokenExpiresAt = credentials.expiresAtMillis

        persistCredentials()
        _isAuthenticated.value = true
    }

    fun signOut() {
        identityId = null
        accessKeyId = null
        secretKey = null
        sessionToken = null
        tokenExpiresAt = 0
        cachedGoogleToken = null
        isBackendRegistered = false

        prefs.edit().clear().apply()
        _isAuthenticated.value = false
        _isFullySetUp.value = false
    }

    // MARK: - Cognito API

    private data class CognitoCredentials(
        val accessKeyId: String,
        val secretKey: String,
        val sessionToken: String,
        val expiresAtMillis: Long
    )

    private fun getCognitoIdentity(googleToken: String): String {
        val url = URL(COGNITO_ENDPOINT)
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.setRequestProperty("Content-Type", "application/x-amz-json-1.1")
        conn.setRequestProperty("X-Amz-Target", "AWSCognitoIdentityService.GetId")
        conn.doOutput = true

        val body = JSONObject().apply {
            put("IdentityPoolId", COGNITO_IDENTITY_POOL_ID)
            put("Logins", JSONObject().apply {
                put("accounts.google.com", googleToken)
            })
        }

        conn.outputStream.use { it.write(body.toString().toByteArray()) }

        val responseCode = conn.responseCode
        if (responseCode !in 200..299) {
            throw ArchiveError.ServerError(responseCode)
        }

        val responseBody = conn.inputStream.bufferedReader().readText()
        val json = JSONObject(responseBody)
        return json.getString("IdentityId")
    }

    private fun getCognitoCredentials(identityId: String, googleToken: String?): CognitoCredentials {
        val url = URL(COGNITO_ENDPOINT)
        val conn = url.openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.setRequestProperty("Content-Type", "application/x-amz-json-1.1")
        conn.setRequestProperty("X-Amz-Target", "AWSCognitoIdentityService.GetCredentialsForIdentity")
        conn.doOutput = true

        val body = JSONObject().apply {
            put("IdentityId", identityId)
            if (googleToken != null) {
                put("Logins", JSONObject().apply {
                    put("accounts.google.com", googleToken)
                })
            }
        }

        conn.outputStream.use { it.write(body.toString().toByteArray()) }

        val responseCode = conn.responseCode
        if (responseCode !in 200..299) {
            throw ArchiveError.ServerError(responseCode)
        }

        val responseBody = conn.inputStream.bufferedReader().readText()
        val json = JSONObject(responseBody)
        val credentials = json.getJSONObject("Credentials")

        return CognitoCredentials(
            accessKeyId = credentials.getString("AccessKeyId"),
            secretKey = credentials.getString("SecretKey"),
            sessionToken = credentials.getString("SessionToken"),
            expiresAtMillis = (credentials.getDouble("Expiration") * 1000).toLong()
        )
    }

    private fun registerWithBackend() {
        val creds = getCredentials() ?: throw ArchiveError.NotAuthenticated
        val identity = identityId ?: throw ArchiveError.NotAuthenticated

        val apiUrl = URL("https://875x1tpmtb.execute-api.us-east-1.amazonaws.com/v1/auth/register")
        val conn = apiUrl.openConnection() as HttpURLConnection
        conn.requestMethod = "POST"
        conn.setRequestProperty("Content-Type", "application/json")
        conn.setRequestProperty("X-Identity-Id", identity)
        conn.doOutput = true

        val body = JSONObject().apply {
            put("source", "android")
        }

        conn.outputStream.use { it.write(body.toString().toByteArray()) }

        // Note: This is unsigned for the register call. For full SigV4 signing,
        // use the Retrofit client with SigV4Interceptor for all other API calls.
        val responseCode = conn.responseCode
        if (responseCode !in 200..299) {
            val errorBody = try { conn.errorStream?.bufferedReader()?.readText() } catch (_: Exception) { "" }
            throw ArchiveError.UploadFailed("Register HTTP $responseCode: $errorBody")
        }
    }

    private fun persistCredentials() {
        prefs.edit()
            .putString(KEY_IDENTITY_ID, identityId)
            .putString(KEY_ACCESS_KEY_ID, accessKeyId)
            .putString(KEY_SECRET_KEY, secretKey)
            .putString(KEY_SESSION_TOKEN, sessionToken)
            .putLong(KEY_TOKEN_EXPIRY, tokenExpiresAt)
            .putString(KEY_GOOGLE_TOKEN, cachedGoogleToken)
            .apply()
    }

    sealed class ArchiveError : Exception() {
        data object NotAuthenticated : ArchiveError()
        data class ServerError(val statusCode: Int) : ArchiveError()
        data class UploadFailed(override val message: String) : ArchiveError()
        data object QuotaExceeded : ArchiveError()
        data object InvalidResponse : ArchiveError()
    }
}
